long-running-tasks/src/test/java/io/streamnative/pulsar/recipes/task/Soak.java

```diff
@@ -0,0 +1,398 @@
+          // Ideally, all but 'received' should be 0
+          log.info(
+              "Received: {}, Inconsistent: {}, MissedResult: {}, MissedCancellation: {}, MissedExcecption: {}, "
+                  + "UnexpectedResult: {}, UnexpectedState: {}",
+              received,
+              incompleteScenarios,
+              missedResults,
+              missedCancellations,
+              missedExceptions,
+              unexpectedResults,
+              unexpectedStates);
+        }
+      }
+      consumer.acknowledge(message);
+    }
+
+    log.info("Soak test completed after {}", Duration.between(start, clock.instant()));
+
+    // close workers
+    workers.stream().forEach(TaskWorker::close);
+
+    assertThat(received).isGreaterThanOrEqualTo(cumulativeTaskTries);
+    assertThat(incompleteScenarios).isEqualTo(0);
+    assertThat(missedResults).isEqualTo(0);
+    assertThat(missedCancellations).isEqualTo(0);
+    assertThat(missedExceptions).isEqualTo(0);
+    assertThat(unexpectedResults).isEqualTo(0);
+    assertThat(unexpectedStates).isEqualTo(0);
+  }
+
+  @SneakyThrows
+  private PulsarClient newPulsarClient(String url) {
+    return PulsarClient.builder().serviceUrl(url).build();
+  }
+
+  @SneakyThrows
+  private TaskWorker newTaskWorker(
+      Process<Task, Result> process,
+      String url,
+      TaskWorkerConfiguration<Task, Result> configuration) {
+    var client = newPulsarClient(url);
+    return TaskWorker.create(client, process, configuration);
+  }
+
+  @ToString
+  @EqualsAndHashCode
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  static class Task {
+    static Task newTask(long id, Duration maxRuntime, List<Outcome> tryOutcomes) {
+      return new Task(maxRuntime.toMillis(), id, tryOutcomes);
+    }
+
+    @Getter private final long maxRuntimeMs;
+    @Getter private final long id;
+    @Getter private final List<Outcome> tryOutcomes;
+
+    @JsonIgnore
+    public Duration getMaxRuntime() {
+      return Duration.ofMillis(maxRuntimeMs);
+    }
+
+    enum Outcome {
+      RESULT,
+      EXCEPTION,
+      TIMEOUT
+    }
+  }
+
+  @AllArgsConstructor(access = AccessLevel.PRIVATE)
+  @ToString
+  static class Result {
+    @Getter private final long elapsedTimeMs;
+    @Getter private final int iteration;
+  }
+
+  @RequiredArgsConstructor
+  static class DummyProcess implements Process<Task, Result> {
+    private final Clock clock;
+    @Getter private final Map<Task, Integer> retryTracker = new HashMap<>();
+
+    @Override
+    public Result apply(@NonNull Task task) throws Exception {
+      int iteration = Optional.ofNullable(retryTracker.get(task)).map(i -> ++i).orElse(0);
+      retryTracker.put(task, iteration);
+      var start = clock.instant();
+      var maxRuntimeMs = task.getMaxRuntime().toMillis();
+      var desiredOutcome = task.getTryOutcomes().get(iteration);
+      switch (desiredOutcome) {
+        case RESULT:
+          Thread.sleep((long) (nextDouble(0.0, 0.9) * maxRuntimeMs));
```

Comment:
What's the reason behind the sleep? Is this just waiting for the task to finish within max runtime? 

1. This isn't accurate when system is under high load.
2. If the task completes quickly you waste time waiting for result. There might be a better way.

Reply:
I think this is just emulating tasks being at least a little long-running.